### PR TITLE
feat: Add Side input proto specifications

### PR DESCRIPTION
**Pull request by kohlisid**
Note: we don't have a key field here because each side input manager is designed to manage only one side input; therefore, the "key" (side input name defined in the pipeline yaml file) is already bind to the side input manager.

### CODE REVIEW

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+message EventTime {
+  // event_time is the time associated with each datum.
+  google.protobuf.Timestamp event_time = 1;
+  // future we can add an enum to represent whether the event time is
+  // based on INGEST_TIME or EVENT_TIME or PROCESSING_TIME
+}
+/**
+ * SideInputResponse represents a datum response element.
```
Comment (jy4096) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:25]: `a datum response element` ??

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
+}
+
+message EventTime {
+  // event_time is the time associated with each datum.
+  google.protobuf.Timestamp event_time = 1;
+  // future we can add an enum to represent whether the event time is
+  // based on INGEST_TIME or EVENT_TIME or PROCESSING_TIME
+}
+/**
```
Comment (jy4096) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:24]: nit: newline between line 23 and 24

**Review by jy4096 — CHANGES_REQUESTED** [review]

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+service UserDefinedSideInput {
+  // SideInputFn applies a function to update the side-input values.
+  rpc RetrieveSideInput(google.protobuf.Empty) returns (SideInputResponse);
+
+  // IsReady is the heartbeat endpoint for gRPC.
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
+}
+
+message EventTime {
+  // event_time is the time associated with each datum.
```
Comment (jy4096) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:19]: `with each datum`

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/sideinput/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package sideinput.v1;
+
+service UserDefinedSideInput {
+  // SideInputFn applies a function to update the side-input values.
```
Comment (jy4096) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:11]:
```suggestion
  // RetrieveSideInput applies a function to retrieve the latest the side-input values.
```

**Review by jy4096 — APPROVED** [review]

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 5]
```diff
@@ -0,0 +1,41 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/sideinput/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package sideinput.v1;
+
+service UserDefinedSideInput {
```
Comment (vigith) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:10]:
```suggestion
UserDefinedSideInput is used to broadcast changes to all the vertices ...... <fill it up>..
service UserDefinedSideInput {
```

Could you please document the expected behavior and use of this proto file? You may link to other documents.

**Review by vigith — CHANGES_REQUESTED** [review]

File: pkg/apis/proto/sideinput/v1/sideinput.proto [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+/**
+ * SideInputResponse represents a response to a given side input retrieval request.
+ */
+message SideInputResponse {
+  // value is the side-input payload
+  bytes value = 1;
+  // event_time represents the timestamp at which value is updated
+  EventTime event_time = 2;
```
Comment (whynowy) [inline pkg/apis/proto/sideinput/v1/sideinput.proto:37]: Do we need this?

The call to `RetrieveSideInput` is supposed to be called by the manager sidecar, I would assume for most of the time, there's no way for the developer to determine the time it got updated, and current time would be used for it.

I guess it is introduced for the purpose of determine if there's any change for the data? I think we should use different approaches for it if it's for that.

Comment (vigith) [reply pkg/apis/proto/sideinput/v1/sideinput.proto:37]: +1, we should remove this as it might cause confusion.

**Review by vigith — APPROVED** [review]

**Review by jy4096 — APPROVED** [review]

**Review by whynowy — APPROVED** [review]

**Pull request merged**